plugins/in_http/http_prot.c

```diff
@@ -289,6 +290,95 @@ static ssize_t parse_payload_json(struct flb_http *ctx, flb_sds_t tag,
     return 0;
 }
 
+static ssize_t parse_payload_urlencoded(struct flb_http *ctx, flb_sds_t tag,
+                                  char *payload, size_t size)
+{
+    struct mk_list *kvs;
+    struct mk_list *head = NULL;
+    struct flb_split_entry *cur = NULL;
+    char **keys = NULL;
+    char **vals = NULL;
+    char *sep;
+    char *start;
+    int idx = 0;
+    int ret = -1;
+    msgpack_packer pck;
+    msgpack_sbuffer sbuf;
+
+
+    /* initialize buffers */
+    msgpack_sbuffer_init(&sbuf);
+    msgpack_packer_init(&pck, &sbuf, msgpack_sbuffer_write);
+
+    kvs = flb_utils_split(payload, '&', -1 );
+    if (kvs == NULL) {
+    	goto split_error;
+    }
+
+    keys = flb_calloc(mk_list_size(kvs), sizeof(char *));
+    if (keys == NULL) {
+    	goto keys_calloc_error;
+    }
+
+    vals = flb_calloc(mk_list_size(kvs), sizeof(char *));
+    if (vals == NULL) {
+    	goto vals_calloc_error;
+    }
+
+    mk_list_foreach(head, kvs) {
+    	cur = mk_list_entry(head, struct flb_split_entry, _head);
+    	if (cur->value[0] == '\n') {
+    	    start = &cur->value[1];
+    	} else {
+    	    start = cur->value;
+	}
+	sep = strchr(start, '=');
+	if (sep == NULL) {
+	    vals[idx] = NULL;
+	    continue;
+	}
+	*sep++ = '\0';
+        
+        keys[idx] = flb_sds_create_len(start, strlen(start));
+	vals[idx] = flb_sds_create_len(sep, strlen(sep));
+
+	flb_sds_trim(keys[idx]);
+	flb_sds_trim(vals[idx]);
+	idx++;
+    }
+
+    msgpack_pack_map(&pck, mk_list_size(kvs));
+    for (idx = 0; idx < mk_list_size(kvs); idx++) {
+        msgpack_pack_str(&pck, flb_sds_len(keys[idx]));
+        msgpack_pack_str_body(&pck, keys[idx], flb_sds_len(keys[idx]));
+
+        msgpack_pack_str(&pck, flb_sds_len(vals[idx]));
+        msgpack_pack_str_body(&pck, vals[idx], flb_sds_len(vals[idx]));
```

Comment:
These lines could be working for ASCII characters. However, when you use non-ASCII characters with URL encoding, they didn't work properly.

```console
% curl -X POST localhost:9880 -H "Content-Type: application/x-www-form-urlencoded" -d "param1=value1&param2=value2"  
% curl -X POST localhost:9880 -H "Content-Type: application/x-www-form-urlencoded" -d "param1=value1&param2=ヤッホー" 
% curl -X POST localhost:9880 -H "Content-Type: application/x-www-form-urlencoded" -d "param1=value1&param2=%E3%83%A4%E3%83%83%E3%83%9B%E3%83%BC" 
% curl -X POST localhost:9880 -H "Content-Type: application/x-www-form-urlencoded" -d "param1=value1&param2=(n‘∀‘)η ﾔｧｰｯﾎｫｰ"                     
% curl -X POST localhost:9880 -H "Content-Type: application/x-www-form-urlencoded" -d "param1=value1&param2=ふるーえんとびっと" 
```

They can be processed as:

```log
[0] http.0: [1678174155.500840000, {"param1"=>"value1", "param2"=>"value2"}]
[0] http.0: [1678174162.973211000, {"param1"=>"value1", "param2"=>"ヤッホー"}]
[0] http.0: [1678174187.523041000, {"param1"=>"value1", "param2"=>"%E3%83%A4%E3%83%83%E3%83%9B%E3%83%BC"}]
[0] http.0: [1678174215.181808000, {"param1"=>"value1", "param2"=>"(n‘∀‘)η ﾔｧｰｯﾎｫｰ"}]
[0] http.0: [1678174238.099359000, {"param1"=>"value1", "param2"=>"ふるーえんとびっと"}]
```

For actual usecases, parameters for application/x-www-form-urlencoded are encoded with URL encoding.